medusa/storage/azure_storage.py

```diff
@@ -1,81 +1,336 @@
+    def download_blobs(self, srcs: t.List[t.Union[Path, str]], dest: t.Union[Path, str]):
+        loop = self._get_or_create_event_loop()
+        loop.run_until_complete(self._download_blobs(srcs, dest))
+
+    async def _download_blobs(self, srcs: t.List[t.Union[Path, str]], dest: t.Union[Path, str]):
+        coros = [self._download_blob(src, dest) for src in map(str, srcs)]
+        await asyncio.gather(*coros)
+
+    async def _download_blob(self, src: str, dest: str):
+
+        # _stat_blob throws if the blob does not exist
+        blob = await self._stat_blob(src)
+        object_key = blob.name
+
+        # we must make sure the blob gets stored under sub-folder (if there is any)
+        # the dest variable only points to the table folder, so we need to add the sub-folder
+        src_path = Path(src)
+        file_path = (
+            "{}/{}/{}".format(dest, src_path.parent.name, src_path.name)
+            if src_path.parent.name.startswith(".")
+            else "{}/{}".format(dest, src_path.name)
+        )
+
+        if blob.size < int(self.config.multi_part_upload_threshold):
+            workers = 1
+        else:
+            workers = int(self.config.concurrent_transfers)
+
+        logging.debug(
+            '[Azure Storage] Downloading with {} workers: {}/{} -> {}'.format(
+                workers, self.config.bucket_name, object_key, file_path
             )
+        )
 
-    def get_object_datetime(self, blob):
+        downloader = await self.azure_container_client.download_blob(
+            blob=object_key,
+            max_concurrency=workers,
+        )
+        await downloader.readinto(open(file_path, "wb"))
+
+    def get_object_datetime(self, blob: AbstractBlob) -> datetime.datetime:
         logging.debug(
             "Blob {} last modification time is {}".format(
-                blob.name, blob.extra["last_modified"]
+                blob.name, blob.last_modified
             )
         )
-        return parser.parse(blob.extra["last_modified"])
+        return blob.last_modified
 
-    def get_cache_path(self, path):
+    def get_cache_path(self, path: str) -> str:
         logging.debug(
             "My cache path is {}".format(path)
         )
         return path
 
-    def upload_blobs(self, srcs, dest):
-        return medusa.storage.azure_blobs_storage.concurrent.upload_blobs(
-            self, srcs, dest, self.bucket,
-            max_workers=self.config.concurrent_transfers,
-            multi_part_upload_threshold=int(self.config.multi_part_upload_threshold)
-        )
-
-    def download_blobs(self, srcs, dest):
-        return medusa.storage.azure_blobs_storage.concurrent.download_blobs(
-            self, srcs, dest, self.bucket,
-            max_workers=self.config.concurrent_transfers,
-            multi_part_upload_threshold=int(self.config.multi_part_upload_threshold)
-        )
+    def _get_or_create_event_loop(self) -> asyncio.AbstractEventLoop:
```

Comment:
**Suggestion:** this seems to be repeated in all the storage classes, maybe we could refactor it to have a single version that all storage classes will use? In the abstract storage possibly?

Reply:
It's indeed repeated, I'll be pushing a different PR fixing this.